L58-python-async-api.md

```diff
@@ -0,0 +1,453 @@
+server.start()
+await server.wait_for_termination()
+```
+
+Client side:
+```Python
+# Python 3.6 and lower
+with grpc.aio.insecure_channel("localhost:50051") as channel:
+    stub = echo_pb2_grpc.EchoStub(channel)
+    response = await stub.Hi(echo_pb2.EchoRequest(message="ping"))
+
+# Or using asynchronous context manager for Python3.7 and up
+async with grpc.aio.insecure_channel("localhost:50051") as channel:
+    stub = echo_pb2_grpc.EchoStub(channel)
+    response = await stub.Hi(echo_pb2.EchoRequest(message="ping"))
+```
+
+### Channel-Side
+
+Changes in `grpc.aio.Channel`:
+* The object returned by `unary_unary`, `unary_stream`, `stream_unary`, and
+  `stream_tream` will be `MultiCallable` objects in `grpc.aio`.
+
+Changes in `grpc.aio.Call`:
+* All methods return `asyncio.Future` object.
+
+Changes for `MultiCallable` classes in `grpc.aio`:
+* `grpc.aio.UnaryUnaryMultiCallable` and `grpc.aio.StreamUnaryMultiCallable`
+  returns `Awaitable` object for `__call__`;
+* `grpc.aio.UnaryUnaryMultiCallable` and `grpc.aio.StreamUnaryMultiCallable`
+  returns `(Awaitable, grpc.aio.Call)` for `with_call`;
+* `grpc.aio.UnaryStreamMultiCallable` and `grpc.aio.StreamStreamMultiCallable`
+  returns an asynchronous generator for `__call__`;
+* `grpc.aio.StreamUnaryMultiCallable` and `grpc.aio.StreamStreamMultiCallable`
+  takes in an asynchronous generator as `request_iterator` argument.
+
+### Server-Side
+
+Changes in `grpc.aio.Server`:
+* `add_generic_rpc_handlers` method accepts tuple of
+  `grpc.aio.GenericRpcHandler`;
+* `wait_for_termination` method is a coroutine;
+* `stop` returns a `asyncio.Event`.
+
+Changes in `grpc.aio.HandlerCallDetails`:
+* `invocation_metadata` method is a coroutine.
+
+Changes in `grpc.aio.GenericRpcHandler`:
+* `service` method takes in `grpc.aio.HandlerCallDetails`;
+* `service` method returns a `grpc.aio.RpcMethodHandler`.
+
+Changes in `grpc.aio.RpcMethodHandler`:
+* All servicer handlers are coroutines;
+* Deserializer / serializer remain normal Python functions.
+
+Changes in `grpc.ServicerContext`:
+* `invocation_metadata` method returns an `asyncio.Future` object;
+* `send_initial_metadata` method is a coroutine and returns an `asyncio.Task`
+  object.
+
+Changes in `grpc.aio.*_*_rpc_method_handler`:
+* Accepts a coroutine;
+* Returns a `grpc.aio.RpcMethodHandler`.
+
+Changes in `grpc.aio.method_handlers_generic_handler`:
+* `method_handlers` is a dictionary that maps method names to corresponding
+  `grpc.aio.RpcMethodHandler`;
+* Returns `grpc.aio.GenericRpcHandler`.
+
+### Interceptor
+
+Changes in `grpc.aio.ServerInterceptor`:
+* Accepts `grpc.aio.HandlerCallDetails`;
+* `continuation` is a function accepts `grpc.aio.HandlerCallDetails` and returns
+  `grpc.aio.RpcMethodHandler`.
+
+Changes in `ClientInterceptor` classes in `grpc.aio`:
+* `continuation` is a coroutine that returns `Awaitable` object.
+
+### Utility Functions
+
+Changes in `grpc.aio.channel_ready_future`:
```

Comment:
What do you think for removing the future suffix, so it will look like `grpc.aio.channel_ready`?

In actual code:
```Python
channel = grpc.aio.insecure_channel("localhost:50051")
await grpc.aio.channel_ready(channel)
```

Reply:
This naming scheme tracks much better for me. `await noun_adjective()` sounds to me like a pretty straightforward way of saying "block until the `noun` is `adjective`".